tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc

```diff
@@ -463,17 +467,28 @@ Status HorizontalLoopFusionImpl::Fuse(
       TF_ASSIGN_OR_RETURN(
           HloInstruction * gep,
           MakeGetTupleElementHlo(hori_fusion_instr, total_output_id++));
-      bitcasts.push_back(computation_->AddInstruction(
-          HloInstruction::CreateBitcast(output->shape(), gep)));
+      // This pass also runs very late. So we can't expect useless
+      // bitcast will be cleaned up.
+      if (ShapeUtil::Equal(gep->shape(), output->shape())) {
+        bitcasts.push_back(gep);
+      } else {
+        bitcasts.push_back(computation_->AddInstruction(
+            HloInstruction::CreateBitcast(output->shape(), gep)));
+      }
     }
     HloInstruction* bitcast_or_tuple =
         (bitcasts.size() == 1) ? bitcasts.at(0)
                                : computation_->AddInstruction(
                                      HloInstruction::CreateTuple(bitcasts));
+    HloComputation* old_computation = fused_instr->fused_instructions_computation();
+    HloModule* module = old_computation->parent();
     TF_RETURN_IF_ERROR(
         computation_->ReplaceInstruction(fused_instr, bitcast_or_tuple));
+    module->RemoveEmbeddedComputation(old_computation);
```

Comment:
Was the computation previously DCEd?

Reply:
Yes, here:
https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/compiler/xla/service/gpu/gpu_compiler.cc;l=600

Note, that DCE is still needed. But it will do less work.